executor/prepared.go

```diff
@@ -0,0 +1,279 @@
+			e.ResultFields = convertResultFields(resultSetNode.GetResultFields())
+		}
+	}
+
+	if e.ID == 0 {
+		e.ID = vars.GetNextPreparedStmtID()
+	}
+	if e.Name != "" {
+		vars.PreparedStmtNameToID[e.Name] = e.ID
+	}
+	vars.PreparedStmts[e.ID] = prepared
+}
+
+// ExecuteExec represents an EXECUTE executor.
+// It executes a prepared statement.
+type ExecuteExec struct {
+	IS        infoschema.InfoSchema
+	Ctx       context.Context
+	Name      string
+	UsingVars []ast.ExprNode
+	ID        uint32
+	StmtExec  Executor
+	OldStmt   stmt.Statement
+}
+
+// Fields implements Executor Fields interface.
+func (e *ExecuteExec) Fields() []*ast.ResultField {
+	// Will never be called.
+	return nil
+}
+
+// Next implements Executor Next interface.
+func (e *ExecuteExec) Next() (*Row, error) {
+	// Will never be called.
+	return nil, nil
+}
+
+// Close implements plan.Plan Close interface.
+func (e *ExecuteExec) Close() error {
+	// Will never be called.
+	return nil
+}
+
+// Build builds a prepared statement into an executor.
+func (e *ExecuteExec) Build() error {
+	vars := variable.GetSessionVars(e.Ctx)
+	if e.Name != "" {
+		e.ID = vars.PreparedStmtNameToID[e.Name]
+	}
+	v := vars.PreparedStmts[e.ID]
+	if v == nil {
+		return ErrStmtNotFound
+	}
+	prepared := v.(*Prepared)
+
+	if len(prepared.Params) != len(e.UsingVars) {
+		return ErrWrongParamCount
+	}
+
+	for i, usingVar := range e.UsingVars {
+		val, err := evaluator.Eval(e.Ctx, usingVar)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		prepared.Params[i].SetValue(val)
+	}
+
+	if optimizer.IsSupported(prepared.Stmt) {
+		if prepared.SchemaVersion != e.IS.SchemaMetaVersion() {
+			// If the schema version has changed we need to prepare it again,
+			// if this time it failed, the real reason for the error is schema changed.
+			err := optimizer.Prepare(e.IS, e.Ctx, prepared.Stmt)
+			if err != nil {
+				return ErrSchemaChanged.Gen("Schema change casued error: %s", err.Error())
+			}
+			prepared.SchemaVersion = e.IS.SchemaMetaVersion()
+		}
+		plan, err := optimizer.Optimize(e.Ctx, prepared.Stmt)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		b := newExecutorBuilder(e.Ctx, e.IS)
+		stmtExec := b.build(plan)
+		if b.err != nil {
+			return errors.Trace(b.err)
+		}
+		e.StmtExec = stmtExec
+	} else {
+		conv := converter.Converter{}
+		oStmt, err := conv.Convert(prepared.Stmt)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		e.OldStmt = oStmt
+	}
+	return nil
+}
+
+// DeallocateExec represent a DEALLOCATE executor.
+type DeallocateExec struct {
+	Name string
+	ctx  context.Context
+}
+
+// Fields implements Executor Fields interface.
+func (e *DeallocateExec) Fields() []*ast.ResultField {
+	return nil
+}
+
+// Next implements Executor Next interface.
+func (e *DeallocateExec) Next() (*Row, error) {
+	vars := variable.GetSessionVars(e.ctx)
+	id, ok := vars.PreparedStmtNameToID[e.Name]
+	if !ok {
+		return nil, ErrStmtNotFound
+	}
+	delete(vars.PreparedStmtNameToID, e.Name)
```

Comment:
Do we need to consider the concurrent issues of PreparedStmtNameToID and PreparedStmts?